### PR TITLE
integrating with Perp88

### PR DESCRIPTION
**Pull request by chentang88158**
- PERP88:

- https://app.perp88.com

- Description
a deriverates project on Polygon chain

1. Tokens

image

2. Pools 
User have to buy PLP to deposit in PLP pool, user can buy PLP with underlying token in PLP 

image

image

3. User Staking
- SuppliedPool query how many user stake in PLP Pool
- RewardedPool query pending reward in USDC

image

### CODE REVIEW

File: apps/workshop/projects/Perp88/helpers/calculation.ts [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+  if (totalRewardToken.isZero()) throw new Error(`_calAPR totalReward is 0`);
+
+  const apr = ethers.utils.formatEther(rewardPricePerYear.mul(e18).div(totalRewardToken).mul(100));
+
+  return apr;
+}
+
+// PRICES
+export async function getPLPPrice(ctx: Context): Promise<TokenPrice> {
```
Comment (witem) [inline apps/workshop/projects/Perp88/helpers/calculation.ts:57]: We added PLP asset tracking to our dashboard, so now you can delete this price logic from your PR

File: apps/workshop/projects/Perp88/modules/PLPStaking.ts [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+  /**
+   * Returns full pool list
+   *
+   * @param context
+   * @returns Pool[]
+   */
+  async fetchPools(ctx: FetchPoolsContext) {
+    const { tokens, ethcall, ethcallProvider, logger } = ctx;
+    const filteredTokens = tokens.filter((i) => i.address.toLowerCase() !== PLP_TOKEN_ADDR);
```
Comment (witem) [inline apps/workshop/projects/Perp88/modules/PLPStaking.ts:56]:
```suggestion
    const filteredTokens = tokens.filter((i) => i.address.toLowerCase() !== PLP_TOKEN_ADDR.toLowerCase());
```

File: apps/workshop/projects/Perp88/modules/PLPStaking.ts [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+    if (tokenAddrs.length !== multiCallRes.length) {
+      throw new Error(`Invalid Multicall Result`);
+    }
+
+    let mapAddrWithResult: Record<string, BigNumber> = {};
+    for (let i = 0; i < Object.values(COMPOSITION_TOKENS).length; i++) {
+      mapAddrWithResult[tokenAddrs[i].toLowerCase()] = multiCallRes[i];
+    }
+
+    const plpToken = {
```
Comment (witem) [inline apps/workshop/projects/Perp88/modules/PLPStaking.ts:78]: You can delete all this static info except the underlying building.

**Review by witem — CHANGES_REQUESTED** [review]
Thanks for your PR. It looks good! I have only few comments:
1. We added PLP to our tracked assets, so now you can delete all related static fields and price calculation logic:
image
image
2. If you keep all these tokens as underlying for PLP, on our dashboard user will see only the breakdown of underlying. It is just to notice. Similar to this:
image
3. Can you please double-check that fetches user positions works correct because I can't get the balance for `0xbb307b0d05378a932f75d5ac2e65d71f48b7b631`:
image

**Comment by chentang88158** [comment]
> 3. Can you please double-check that fetches user positions works correct because I can't get the balance for `0xbb307b0d05378a932f75d5ac2e65d71f48b7b631`:
>    image

>  This user holding plp in wallet, balance will displayed only wallet that staked in our pool only

**Comment by witem** [comment]
@chentang88158 
> This user holding plp in wallet, balance will display only wallet that staked in our pool only

Can you share a wallet that has staked in the pool?

> How often is the new price updated?, since our perp88 website plp is 1.13, but in your graph is 1.1.6?

In prod, the price will be updated every 15 minutes. For the sandbox, there is a long leaving cache(around one day)

> Do you have any method to force fetch new price ?

We don't have any public methods for it

File: apps/workshop/projects/Perp88/helpers/calculation.ts [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+  const plpStakingContract = new ethcall.Contract(PLP_STAKING_ADDR, PLPStakingAbi);
+
+  const [rewardRate, totalShare]: BigNumber[] = await ethcallProvider.all([
+    rewardContract.rewardRate(),
+    plpStakingContract.calculateTotalShare(PLP_STAKING_REVENUE_ADDR),
+  ]);
+
+  const rewardRateE18 = rewardRate.mul(10 ** 12); //usdc is 1e6
+
+  const price = await getTokenOraclePrice(ctx, COMPOSITION_TOKENS.USDC);
```
Comment (witem) [inline apps/workshop/projects/Perp88/helpers/calculation.ts:46]: We're also providing a price for the USDC token, so you don't need to get it from the oracle

Comment (chentang88158) [reply apps/workshop/projects/Perp88/helpers/calculation.ts:46]: is provided price from chainlink?

**Comment by chentang88158** [comment]
> @chentang88158
> 
> > This user holding plp in wallet, balance will display only wallet that staked in our pool only
> 
> Can you share a wallet that has staked in the pool?
> 0x841f4b2c13223fce62cace03f4c9b4ecadcff1ee

**Review by witem — DISMISSED** [review]

File: apps/workshop/projects/Perp88/helpers/calculation.ts (hunk 4 above)
Comment (witem) [reply apps/workshop/projects/Perp88/helpers/calculation.ts]: We are getting prices from different places and calculating the average.

**Comment by witem** [comment]
@chentang88158 there is problems with build: https://github.com/defiyield-app/defiyield-self-integrations/actions/runs/3992787153/jobs/6856241401 :

```
projects/Perp88/index.ts → dist/Perp88.js...
(!) `this` has been rewritten to `undefined`
https://rollupjs.org/guide/en/#error-this-is-undefined
../../.yarn/cache/ethers-npm-5.7.2-eba7d781ee-b7c08cf3e2.zip/node_modules/ethers/lib/utils.js
1: "use strict";
2: var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
   ^
3:     if (k2 === undefined) k2 = k;
4:     Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
...and 5 other occurrences
../../.yarn/cache/@ethersproject-abstract-signer-npm-5.7.0-f61d0a970e-a823dac9cf.zip/node_modules/@ethersproject/abstract-signer/lib.esm/index.js
1: "use strict";
2: var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
   ^
3:     function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
4:     return new (P || (P = Promise))(function (resolve, reject) {
...and 1 other occurrence
../../.yarn/cache/@ethersproject-wallet-npm-5.7.0-f553c726b7-a4009bf733.zip/node_modules/@ethersproject/wallet/lib.esm/index.js
1: "use strict";
2: var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
   ^
3:     function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
4:     return new (P || (P = Promise))(function (resolve, reject) {
...and 1 other occurrence

...and 17 other files
[!] Error: 'default' is not exported by ../../.yarn/cache/bn.js-npm-5.2.1-dc952b1965-3dd8c8d380.zip/node_modules/bn.js/lib/bn.js, imported by ../../.yarn/cache/@ethersproject-bignumber-npm-5.7.0-cd761880ac-8c9a134b76.zip/node_modules/@ethersproject/bignumber/lib.esm/bignumber.js
https://rollupjs.org/guide/en/#error-name-is-not-exported-by-module
../../.yarn/cache/@ethersproject-bignumber-npm-5.7.0-cd761880ac-8c9a134b76.zip/node_modules/@ethersproject/bignumber/lib.esm/bignumber.js (9:0)
 7:  *
 8:  */
 9: import _BN from "bn.js";
    ^
10: var BN = _BN.BN;
11: import { hexlify, isBytes, isHexString } from "@ethersproject/bytes";
Error: 'default' is not exported by ../../.yarn/cache/bn.js-npm-5.2.1-dc952b1965-3dd8c8d380.zip/node_modules/bn.js/lib/bn.js, imported by ../../.yarn/cache/@ethersproject-bignumber-npm-5.7.0-cd761880ac-8c9a134b76.zip/node_modules/@ethersproject/bignumber/lib.esm/bignumber.js
    at error (/home/runner/work/defiyield-self-integrations/defiyield-self-integrations/.yarn/cache/rollup-npm-2.79.1-94e707a9a3-6a2bf167b3.zip/node_modules/rollup/dist/shared/rollup.js:198:30)
    at Module.error (/home/runner/work/defiyield-self-integrations/defiyield-self-integrations/.yarn/cache/rollup-npm-2.79.1-94e707a9a3-6a2bf167b3.zip/node_modules/rollup/dist/shared/rollup.js:12560:16)
    at Module.traceVariable (/home/runner/work/defiyield-self-integrations/defiyield-self-integrations/.yarn/cache/rollup-npm-2.79.1-94e707a9a3-6a2bf167b3.zip/node_modules/rollup/dist/shared/rollup.js:12919:29)
    at ModuleScope.findVariable (/home/runner/work/defiyield-self-integrations/defiyield-self-integrations/.yarn/cache/rollup-npm-2.79.1-94e707a9a3-6a2bf167b3.zip/node_modules/rollup/dist/shared/rollup.js:11571:39)
    at MemberExpression.bind (/home/runner/work/defiyield-self-integrations/defiyield-self-integrations/.yarn/cache/rollup-npm-2.79.1-94e707a9a3-6a2bf167b3.zip/node_modules/rollup/dist/shared/rollup.js:8732:49)
    at VariableDeclarator.bind (/home/runner/work/defiyield-self-integrations/defiyield-self-integrations/.yarn/cache/rollup-npm-2.79.1-94e707a9a3-6a2bf167b3.zip/node_modules/rollup/dist/shared/rollup.js:5400:23)
    at VariableDeclaration.bind (/home/runner/work/defiyield-self-integrations/defiyield-self-integrations/.yarn/cache/rollup-npm-2.79.1-94e707a9a3-6a2bf167b3.zip/node_modules/rollup/dist/shared/rollup.js:5396:73)
    at Program.bind (/home/runner/work/defiyield-self-integrations/defiyield-self-integrations/.yarn/cache/rollup-npm-2.79.1-94e707a9a3-6a2bf167b3.zip/node_modules/rollup/dist/shared/rollup.js:5396:73)
    at Module.bindReferences (/home/runner/work/defiyield-self-integrations/defiyield-self-integrations/.yarn/cache/rollup-npm-2.79.1-94e707a9a3-6a2bf167b3.zip/node_modules/rollup/dist/shared/rollup.js:12556:18)
    at Graph.sortModules (/home/runner/work/defiyield-self-integrations/defiyield-self-integrations/.yarn/cache/rollup-npm-2.79.1-94e707a9a3-6a2bf167b3.zip/node_modules/rollup/dist/shared/rollup.js:23193:20)
```

**Comment by witem** [comment]
@chentang88158, you can take a look at this comment https://github.com/defiyield-app/defiyield-self-integrations/pull/12#pullrequestreview-1269772280
it's about similar problems with the build. In general, you shouldn't use any module imports expect relative files, `@defiyield` modules, and `import type`. To use `ethers` or other modules, you should use injected ctx

**Comment by chentang88158** [comment]
> @chentang88158, you can take a look at this comment [#12 (review)](https://github.com/defiyield-app/defiyield-self-integrations/pull/12#pullrequestreview-1269772280) it's about similar problems with the build. In general, you shouldn't use any module imports expect relative files, `@defiyield` modules, and `import type`. To use `ethers` or other modules, you should use injected ctx

Done. Thank you

**Comment by witem** [comment]
@chentang88158 thanks for the updates. Also, there is a problem with failed tests for perp88:
image

**Comment by chentang88158** [comment]
> @chentang88158 thanks for the updates. Also, there is a problem with failed tests for perp88: image

sorry for that, but it seem like `createMockProvider` has no method in real `ethers` use, what I do is export `ethers` from `src/provider` and `createMockContext.ts` use these things. you can comment again for fixing these logics.

**Comment by witem** [comment]
Moved code to https://github.com/defiyield-app/defiyield-self-integrations/pull/13 with linter and prettier fixes

**Pull request merged**